tests/system_test.py

```diff
@@ -317,19 +322,55 @@ def teardown(self):
 
         def error(msg):
             with open(self.outfile + '.out') as f:
-                raise RuntimeError("Process %s error: %s\n%s\n%s\n>>>>\n%s<<<<" % (
-                    self.pid, msg, ' '.join(self.args),
+                raise RuntimeError("Process %s (name=%s) error: %s\n%s\n%s\n>>>>\n%s<<<<" % (
+                    self.pid, self.name, msg, ' '.join(self.args),
                     self.outfile + '.cmd', f.read()))
 
+        # If the process is already dead, this might mean that the process has crashed (For example: a router
+        # crash). self.poll() will return a non None value if the process is already gone.
+        # A None value indicates that the process hasn’t terminated yet.
+        # A negative value -N indicates that the child was terminated by signal N (POSIX only).
+        # A positive value indicates a fail.
         status = self.poll()
-        if status is None:      # Still running
+
+        if status is None:
+            # If status is None, it means that the process is still running
+            self.actual = Process.RUNNING
+        elif status > 0:
+            self.actual = Process.EXIT_FAIL
+        else:
+            self.actual = status
+
+        if status is None:
+            # The process is still running.
+            # Call process.terminate() which will try to stop the child by sending it a SIGTERM
+            # but is not guaranteed to stop it
             self.terminate()
-            if self.expect is not None and self.expect != Process.RUNNING:
-                error("still running")
-            self.expect = 0     # Expect clean exit after terminate
-            status = self.wait()
-        if self.expect is not None and self.expect != status:
```

Comment:
This is sufficiently complicated logic that I am tempted to suggest there should be a test for it :P

(I'm actually almost serious; writing tests for tests is nonsense, but writing tests for test helpers like this is IMO not otter nonsense.)

Reply:
The existing tests are testing the expect and actual (<0, 0, >0) cases. As far as the process termination part is concerned, every teardown() will go thru it, so bugs in this function will bubble up immediately.